lib/acl-checker.js

```diff
@@ -1,172 +1,143 @@
 
 const DEFAULT_ACL_SUFFIX = '.acl'
 
+// An ACLChecker exposes the permissions on a specific resource
 class ACLChecker {
-  constructor (options = {}) {
-    this.debug = options.debug || console.log.bind(console)
+  constructor (resource, options = {}) {
+    this.resource = resource
+    this.host = options.host
+    this.origin = options.origin
     this.fetch = options.fetch
     this.strictOrigin = options.strictOrigin
     this.suffix = options.suffix || DEFAULT_ACL_SUFFIX
   }
 
-  can (user, mode, resource, callback, options = {}) {
-    const debug = this.debug
-    debug('Can ' + (user || 'an agent') + ' ' + mode + ' ' + resource + '?')
-    var accessType = 'accessTo'
-    var possibleACLs = ACLChecker.possibleACLs(resource, this.suffix)
+  // Returns a fulfilled promise when the user can access the resource
+  // in the given mode, or rejects with an HTTP error otherwise
+  can (user, mode) {
+    debug(`Can ${user || 'an agent'} ${mode} ${this.resource}?`)
     // If this is an ACL, Control mode must be present for any operations
-    if (this.isAcl(resource)) {
+    if (this.isAcl(this.resource)) {
       mode = 'Control'
     }
-    var self = this
-    async.eachSeries(
-      possibleACLs,
-
-      // Looks for ACL, if found, looks for a rule
-      function tryAcl (acl, next) {
-        debug('Check if acl exist: ' + acl)
-        // Let's see if there is a file..
-        self.fetch(acl, function (err, graph) {
-          if (err || !graph || graph.length === 0) {
-            if (err) debug('Error: ' + err)
-            accessType = 'defaultForNew'
-            return next()
-          }
-          self.checkAccess(
-            graph, // The ACL graph
-            user, // The webId of the user
-            mode, // Read/Write/Append
-            resource, // The resource we want to access
-            accessType, // accessTo or defaultForNew
-            acl, // The current Acl file!
-            (err) => { return next(!err || err) },
-            options
-          )
-        })
-      },
-      function handleNoAccess (err) {
-        if (err === false || err === null) {
-          debug('No ACL resource found - access not allowed')
-          err = new Error('No Access Control Policy found')
-        }
-        if (err === true) {
-          debug('ACL policy found')
-          err = null
-        }
-        if (err) {
-          debug('Error: ' + err.message)
-          if (!user || user.length === 0) {
-            debug('Authentication required')
-            err.status = 401
-            err.message = 'Access to ' + resource + ' requires authorization'
+
+    // Obtain the permission set for the resource
+    if (!this._permissionSet) {
+      this._permissionSet = this.getNearestACL()
+        .then(acl => this.getPermissionSet(acl))
+    }
+
+    // Check the resource's permissions
+    return this._permissionSet.then(acls => this.checkAccess(acls, user, mode))
+    .catch(err => {
```

Comment:
Done in bda742636e346fff02f6f082b10d37bbd524013b, let me know if more indentation needs adjustments.